### PR TITLE
Created PDF report generation scripts

### PR DESCRIPTION
**Pull request by ebradley12**
This pull request refers to ticket #38 

This pull request introduces a script to generate a daily sales report in PDF format, summarising key sales metrics such as total transactions, total sales, top genres, top artists, and top regions. The report is generated for the previous day, uploaded to an S3 bucket, and emailed to recipients using AWS services.

### CODE REVIEW

File: reports/report_generator.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            "top_genre": top_genres[0][0] if top_genres else "N/A",
+            "top_artist": top_artists[0][0] if top_artists else "N/A"
+        }
+    except Exception as e:
+        logging.error("Error querying sales data: %s", e)
+        return {}
+
+
+def generate_pdf(data, output_file):
+    """Generate a PDF report containing sales data."""
```
Comment (emily-curtis) [inline reports/report_generator.py:107]: Could we have type hints?

File: reports/report_generator.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                for line in content:
+                    pdf.cell(200, 10, txt=line, ln=True)
+                pdf.ln(5)
+
+        pdf.output(output_file)
+        logging.info(f"PDF report generated at {output_file}")
+    except Exception as e:
+        logging.error("Error generating PDF: %s", e)
+        raise
+
```
Comment (emily-curtis) [inline reports/report_generator.py:138]: Great use of logging!!

File: reports/report_generator.py (hunk 1 above)
Comment (ebradley12) [reply reports/report_generator.py:107]: Yes ofc how could I forget!

**Comment by lukieh2014** [comment]
we could add a couple graphs to the report, to visualise the data ?

File: reports/graphs.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+"""
+
+import io
+import logging
+import matplotlib.pyplot as plt
+
+
+def generate_bar_chart(data: list, title: str, xlabel: str, ylabel: str) -> io.BytesIO:
+    """
+    Generate a bar chart.
```
Comment (GabrielNsiah) [inline reports/graphs.py:13]: Could this docstring be more descriptive?

File: docker/pipeline_dockerfile [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+COPY requirements.txt .
+
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY . .
+
+RUN pip install python-dotenv
+
+CMD ["python", "etl.py"]
```
Comment (bentrzcinski) [inline docker/pipeline_dockerfile:13]: Is this related to the pull request? We already have a dockerfile for this in pipline (and this needs to be a lambda dockerfile)

File: reports/emailer.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def send_email_with_attachment(pdf_file: str, recipient_emails: list,
+                               subject: str, body_text: str) -> None:
+    """
+    Send an email with a PDF attachment to multiple recipients using AWS SES.
+    """
+    try:
+        ses_client = boto3.client(
+            'ses', region_name=environ.get('AWS_REGION', 'us-east-1'))
```
Comment (bentrzcinski) [inline reports/emailer.py:35]:
```suggestion
            'ses', region_name=environ.get('AWS_REGION', 'eu-west-2'))
```

File: reports/emailer.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                }
+            )
+            logging.info("Email sent successfully to %s: %s",
+                         recipient_email, response)
+    except ClientError as e:
+        logging.error("Failed to send email: %s", e)
+        raise
+    except Exception as e:
+        logging.error("Error sending emails: %s", e)
+        raise
```
Comment (bentrzcinski) [inline reports/emailer.py:69]: If the custom error messages are the same, these exceptions can be caught in the same line: 
```suggestion
    except (ClientError, Exception) as e:
        logging.error("Failed to send email: %s", e)
        raise
```

File: reports/graphs.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+"""
+This script provides functions for generating visualisations as part of the PDF Report.
+
+"""
```
Comment (bentrzcinski) [inline reports/graphs.py:4]:
```suggestion
"""
This script provides functions for generating visualisations as part of the PDF Report.
"""
```

File: reports/graphs.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        plt.title(title, fontsize=16)
+        plt.xlabel(xlabel, fontsize=14)
+        plt.ylabel(ylabel, fontsize=14)
+        plt.tight_layout()
+        img_buffer = io.BytesIO()
+        plt.savefig(img_buffer, format='png')
+        plt.close()
+        img_buffer.seek(0)
+        return img_buffer
+    except Exception as e:
```
Comment (bentrzcinski) [inline reports/graphs.py:30]: Could we make this a more specific exception?

File: reports/lambda_handler.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+
+def lambda_handler(event: dict, context=None) -> dict:
+    """
+    AWS Lambda function to generate a PDF report, upload it to S3, and email it to subscribers.
+    """
+    try:
+        logging.basicConfig(
+            format="%(asctime)s - %(levelname)s - %(message)s", level=logging.INFO
+        )
```
Comment (bentrzcinski) [inline reports/lambda_handler.py:22]: Could logging config be done in its own function?

File: reports/lambda_handler.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        else:
+            logging.info("No subscribers opted in for the PDF report.")
+
+        return {
+            "statusCode": 200,
+            "body": "Report generated and emailed successfully."
+        }
+    except ValueError as ve:
+        logging.error("Validation error: %s", ve)
+        return {"statusCode": 400, "body": "Validation Error: %s" % ve}
```
Comment (bentrzcinski) [inline reports/lambda_handler.py:51]: Can we only get 200 or 400 codes? What if the server can't be found?

File: reports/queries.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+            port=environ["DB_PORT"],
+            user=environ["DB_USER"],
+            password=environ["DB_PASSWORD"],
+            database=environ["DB_NAME"]
+        )
+        logging.info("Connected successfully.")
+        return connection
+
+    except psycopg2.OperationalError:
+        logging.warning("The database %s doesn't exist", environ["DB_NAME"])
```
Comment (bentrzcinski) [inline reports/queries.py:42]: What is DB_NAME isn't defined? We don't want the error handling throwing an error

File: reports/report_generator.py [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+    Add the Top Genres by Revenue graph to the PDF report.
+    """
+    try:
+        buffer = generate_bar_chart(current_data.get(
+            "top_genres", []), "Top Genres by Revenue", "Revenue ($)", "")
+        with tempfile.NamedTemporaryFile(suffix=".png", delete=True) as tmp_file:
+            tmp_file.write(buffer.read())
+            pdf.image(tmp_file.name, x=30, y=pdf.get_y(), w=150)
+            pdf.ln(20)
+    except ValueError as e:
```
Comment (bentrzcinski) [inline reports/report_generator.py:109]: Could we have a more specific exception type?

File: reports/report_generator.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+    Add the Top Artists by Revenue graph to the PDF report.
+    """
+    try:
+        buffer = generate_bar_chart(current_data.get(
+            "top_artists", []), "Top Artists by Revenue", "Revenue ($)", "")
+        with tempfile.NamedTemporaryFile(suffix=".png", delete=True) as tmp_file:
+            tmp_file.write(buffer.read())
+            pdf.image(tmp_file.name, x=30, y=pdf.get_y(), w=150)
+            pdf.ln(20)
+    except ValueError as e:
```
Comment (bentrzcinski) [inline reports/report_generator.py:127]: more specific exception type?

File: reports/report_generator.py [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+    Add the Top Countries by Revenue graph to the PDF report.
+    """
+    try:
+        buffer = generate_bar_chart(current_data.get(
+            "top_regions", []), "Top Countries by Revenue", "Revenue ($)", "")
+        with tempfile.NamedTemporaryFile(suffix=".png", delete=True) as tmp_file:
+            tmp_file.write(buffer.read())
+            pdf.image(tmp_file.name, x=30, y=pdf.get_y(), w=150)
+            pdf.ln(20)
+    except ValueError as e:
```
Comment (bentrzcinski) [inline reports/report_generator.py:145]: Could we have a more specific exception type here?

File: reports/report_generator.py [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+        add_top_artists_graph(pdf, current_data)
+        pdf.ln(100)
+        add_top_countries_graph(pdf, current_data)
+        pdf.add_page()
+        add_sales_over_time_graph(pdf, current_data)
+
+        pdf.output(output_file)
+        logging.info(f"PDF report generated at {output_file}")
+
+    except Exception as e:
```
Comment (bentrzcinski) [inline reports/report_generator.py:196]: How about a specific exception type here for likely errors?

File: reports/report_generator.py [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+        current_data = query_sales_data(formatted_yesterday)
+        previous_data = query_sales_data(formatted_day_before_yesterday)
+
+        pdf_file = f"/tmp/daily_sales_report_{formatted_yesterday}.pdf"
+        format_pdf(current_data, previous_data,
+                   pdf_file, formatted_yesterday)
+
+        logging.info(f"PDF report generated: {pdf_file}")
+        return pdf_file
+    except Exception as e:
```
Comment (bentrzcinski) [inline reports/report_generator.py:222]: Could we have a more specific exception type here?

File: reports/report_generator.py [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+        logging.error(f"Failed to generate PDF report: {e}")
+        raise
+
+
+if __name__ == "__main__":
+    pdf_file = generate_daily_report()
```
Comment (bentrzcinski) [inline reports/report_generator.py:228]: Do we want to save this to a variable here?

**Review by bentrzcinski — CHANGES_REQUESTED** [review]
Really nice code, I particularly like the separate queries script.

**Review by bentrzcinski — APPROVED** [review]
Looks good! Nice changes

**Pull request merged**